proposals/dom_create_port.md

```diff
@@ -0,0 +1,206 @@
+
+Script that executes in the main world:
+```js
+function getFoo() { ... }
+function getBar() { ... }
+// Other code dispatches 'foochanged' and 'barchanged' events.
+```
+
+Content script:
+```js
+// This function will execute in the main world.
+function setUpMainWorld(port) {
+  // This responds to a message received from the content script.
+  port.onMessage = (message) => {
+    if (message == 'getFoo') {
+      const foo = getFoo();
+      port.sendMessage({foo});
+    } else if (message == 'getBar') {
+      const bar = getBar();
+      port.sendMessage({bar});
+    }
+  };
+
+  // These notify the content script of changes.
+  addEventListener('foochanged', () => {
+    port.sendMessage({foo: getFoo()});
+  });
+  addEventListener('barchanged', () => {
+    port.sendMessage({bar: getBar()});
+  });
+}
+
+// The rest of this code executes in the content script world.
+const mainWorldPort = browser.dom.createPort({world: 'MAIN'});
+
+mainWorldPort.onMessage = (message) => {
+  if (message.foo) {
+    updateFoo(message.foo);
+  } else if (message.bar) {
+    updateBar(message.bar);
+  }
+};
+
+browser.dom.execute(
+    {
+      func: setUpMainWorld,
+      args: mainWorldPort,
+    });
+
+function fetchFoo() {
+  mainWorldPort.sendMessage('getFoo');
+}
+
+function fetchBar() {
+  mainWorldPort.sendMessage('getBar');
+}
+```
+
+##### Message serialization
+
+Message arguments are serialized and deserialized using the Structured Cloning
+algorithm.  This allows for more flexibility than simply JSON conversion.
+Message ports cannot pass other message ports in arguments.
+
+### New Permissions
+
+There are no new permissions for this capability.  This does not allow any new
+data access, since it is only accessible once the extension has already
+injected a script into the document.  Extensions can also already interact with
+the main world of the document through either appending script tags or directly
+injecting with registered content or user scripts, or the
+scripting.executeScript() method.
+
+### Manifest File Changes
+
+There are no necessary manifest changes.
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+This does not result in any new data being exposed.
+
+### Abuse Mitigations
+
+This doesn't enable any new data access.  To reduce risk of cross-world
+contamination, extensions must specify the world with which they want to
+communicate, and all arguments are copied rather than directly shared.
+
+### Additional Security Considerations
+
+N/A.
+
+## Alternatives
+
+### Existing Workarounds
+
+Today, to communicate between JS worlds, extensions can use custom events as
+described in the [content script
+documentation](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#host-page-communication).
+This is fragile and hacky, and can lean to leaking more data to the embedding
+page.
+
+### Open Web API
+
+The open web is unaware of the concept of multiple Javascript worlds, so this
```

Comment:
Should we rename `MessagePort` to `SyncMessagePort`? That would both address the naming called out in the previous comment and describe what is different about this interface.

Reply:
> Should we rename MessagePort to SyncMessagePort?

Rather `MessagePortSync` because `Sync` at the end is the established pattern in the web platform (FileEntrySync) and nodejs (fs.readFileSync), whereas `Sync` at the beginning indicates it being the primary subject e.g. SyncEvent is for synchronization events.

Reply:
> we cannot use MessagePort because it is async

Indeed, so maybe just make a new simpler API that doesn't use the concept of ports, but just establishes one secure synchronous channel.